test/07_versioned-releases.ts

```diff
@@ -774,4 +1398,209 @@ describe('draft releases', () => {
+
+	[
+		['1.2.3+build1', '', 'build1', 0],
+		['1.2.3+build2', '', 'build2', 1],
+		['1.2.4-beta1+build1', 'beta1', 'build1', 0],
+	].forEach(([semver, prerelease, build, resultingRevision]) => {
+		let release: AnyObject;
+		it(`should support draft releases with semver build metadata parts: ${semver}`, async () => {
+			release = (
+				await pineUser
+					.post({
+						resource: 'release',
+						body: {
+							...newReleaseBody,
+							commit: randomUUID(),
+							semver,
+							is_final: false,
+						},
+					})
+					.expect(201)
+			).body;
+
+			expect(release).to.have.property('semver', semver);
+			expect(release).to.have.property('semver_prerelease', prerelease);
+			expect(release).to.have.property('semver_build', build);
+			expect(release).to.have.property('revision', null);
+			expect(release).to.have.property('is_final', false);
+			expectCorrectReleaseComputedTerms(release);
+		});
+
+		it(`should be able to finalize draft releases with semver build metadata parts: ${semver}`, async () => {
+			await pineUser
+				.patch({
+					resource: 'release',
+					id: release.id,
+					body: {
+						is_final: true,
+					},
+				})
+				.expect(200);
+			const { body: updatedRelease } = await pineUser
+				.get({
+					resource: 'release',
+					id: release.id,
+					options: {
+						$select: releaseComputedTermsRequiredFields,
+					},
+				})
+				.expect(200);
+
+			expect(updatedRelease).to.have.property('semver', semver);
+			expect(updatedRelease).to.have.property('semver_build', build);
+			expect(updatedRelease).to.have.property('revision', resultingRevision);
+			expect(updatedRelease).to.have.property('is_final', true);
+			expectCorrectReleaseComputedTerms(updatedRelease);
+		});
+	});
+
+	[
+		['1.2.3+rev1', 'rev1', 1, false],
+		['1.2.3+build1.rev1', 'build1.rev1', 1, false],
+		['1.2.3+rev1.build1', 'rev1.build1', 1, false],
+		['1.2.3+rev2', 'rev2', 2, true],
+		['1.2.3+build3.rev3', 'build3.rev3', 3, true],
+		['1.2.3+rev4.build4', 'rev4.build4', 4, true],
+		['1.2.4+rev0', 'rev0', 0, true],
+	].forEach(([semver, build, resultingRevision, canFinalize]) => {
+		let release: AnyObject;
+		it(`should support draft releases with user provided revN semver build metadata parts: ${semver}`, async () => {
+			release = (
+				await pineUser
+					.post({
+						resource: 'release',
+						body: {
+							...newReleaseBody,
+							commit: randomUUID(),
+							semver,
+							is_final: false,
+						},
+					})
+					.expect(201)
+			).body;
+
+			expect(release).to.have.property('semver', semver);
+			expect(release).to.have.property('semver_build', build);
+			expect(release).to.have.property('revision', null);
+			expect(release).to.have.property('is_final', false);
+			expectCorrectReleaseComputedTerms(release);
+		});
+
+		if (!canFinalize) {
+			it(`should not be able to finalize draft releases with user provided revN semver build metadata parts when the generated revision is different: ${semver}`, async () => {
```

Comment:
When we added the rest semver fields that had 0 as the default value we didn't notice any performance issue.
I realize that NVARCHAR is might not have the same const as an INTEGER field, but on the other hand PG11+ claims that it will only add the default values lazily, so that the ALTER TABLE doesn't cause a full table lock.
See: https://www.postgresql.org/docs/current/ddl-alter.html#DDL-ALTER-ADDING-A-COLUMN

Reply:
So the requirement is: every release that is finalized and has (user provided) build metadata that includes a `rev\d+` part, should have an auto-generated revision that matches that (user provided) `rev\d+` build metadata part.
In other words: If a user creates a finalized release or finalize a draft release that has (user provided) build metadata that includes a `rev\d+` part, the request should only success iff the (user provided) `rev\d+` part matches the revision that that the API will generate.
I'm not sure how make that clearer in the `it()`, but I will add some extra test inputs with a comment of why this is expected to fail.

Reply:
Thanks for adding the logic on the test cases in the array